lib/beaker/host/unix/file.rb

```diff
@@ -1,8 +1,8 @@
 module Unix::File
   include Beaker::CommandFactory
 
-  def tmpfile(name = '')
-    execute("mktemp -t #{name}.XXXXXX")
+  def tmpfile(name = '', extension = nil)
+    execute("mktemp -t #{name}.XXXXXX#{extension}")
```

Comment:
This is not portable, for example the FreeBSD man page for `mktemp(1)` says:

> The template may be any  file name with some number of ‘Xs’ appended to it, for example `/tmp/temp.XXXX`.  The trailing ‘Xs’ are replaced with the current process number and/or a unique letter combination.

And indeed, it does not do the right thing on FreeBSD:

```sh-session
romain@zappy ~ % mktemp -t foo.XXXX.gz
/tmp/foo.XXXX.gz.uAo9BCUu
```

I am not sure there is something guaranteed to be portable given `mktemp` is not defined by POSIX.  The GNU/Linux and FreeBSD man pages at least contain references to function that support this:

> The `mkstemps()` and `mkostemps()` functions act the same as `mkstemp()` and `mkostemp()` respectively, except they permit a suffix to exist in the template.  The template should be of the form `/tmp/tmpXXXXXXsuffix`.  The `mkstemps()` and `mkostemps()` function are told the length of the suffix string.

Reply:
Thanks for that insight. I wonder what would be best to do here. Given Windows does not implement it either, would it be OK to mention it as a potential compatibility issue in a param documentation? I mean, in the end it still gives you a unique filename.

Reply:
That looks great!

Reply:
Wait wait wait, why don't we use `Tempfile` from stdlib which [already has support for this](https://ruby-doc.org/stdlib-2.7.0/libdoc/tempfile/rdoc/Tempfile.html#method-c-new):

> The basename parameter is used to determine the name of the temporary file. You can either pass a String or an Array with 2 String elements. In the former form, the temporary file’s base name will begin with the given string. In the latter form, the temporary file’s base name will begin with the array’s first element, and end with the second element. For example:

```ruby
file = Tempfile.new('hello')
file.path  # => something like: "/tmp/hello2843-8392-92849382--0"
         
# Use the Array form to enforce an extension in the filename:
file = Tempfile.new(['hello', '.jpg'])
file.path  # => something like: "/tmp/hello2843-8392-92849382--0.jpg"
```

Reply:
Because the Ruby part is executed on the host, not the target. You may need to do some wacky things to get Ruby to execute on the target host and in beaker we can't assume it's present since Puppet may not have been bootstrapped.